main/src/ca/uwaterloo/flix/language/phase/Dependencies.scala

```diff
@@ -1,88 +1,719 @@
+    sctx.deps.put((src.sp1.source.input, dst.sp1.source.input), ())
+  }
 
-      /**
-        * Adds a dependency `src -> dst` signifying that if `src` changes then `dst` must be recomputed.
-        */
-      private def addDependency(src: SourceLocation, dst: SourceLocation): Unit = {
-        deps += (src.sp1.source.input, dst.sp1.source.input)
-      }
+  /**
+    * Returns the dependencies of the given type.
+    */
+  private def visitType(tpe: Type)(implicit sctx: SharedContext): Unit = tpe match {
+    case Type.Apply(tpe1, tpe2, _) =>
+      visitType(tpe1)
+      visitType(tpe2)
+    case Type.Alias(cst, args, _, loc) =>
+      addDependency(cst.loc, loc)
+      args.foreach(visitType)
+    case Type.AssocType(_, arg, _, _) => visitType(arg)
+    case Type.JvmToType(tpe, _) => visitType(tpe)
+    case Type.JvmToEff(tpe, _) => visitType(tpe)
+    case Type.UnresolvedJvmType(_, _) => ()
+    case Type.Cst(TypeConstructor.Enum(sym, _), loc) => addDependency(sym.loc, loc)
+    case Type.Cst(TypeConstructor.Struct(sym, _), loc) => addDependency(sym.loc, loc)
+  }
 
-      override def consumeAssocTypeSymUse(symUse: SymUse.AssocTypeSymUse): Unit = {
-        addDependency(symUse.sym.loc, symUse.loc)
-      }
+  /**
+    * Returns the dependencies of the given SymUse.
+    */
+  private def visitSymUse(use: SymUse)(implicit sctx: SharedContext): Unit = use match {
+    case SymUse.AssocTypeSymUse(sym, loc) => addDependency(sym.loc, loc)
+    case SymUse.CaseSymUse(sym, loc) => addDependency(sym.loc, loc)
+    case SymUse.DefSymUse(sym, loc) => addDependency(sym.loc, loc)
+    case SymUse.EffectSymUse(sym, qname) => addDependency(sym.loc, qname.loc)
+    case SymUse.OpSymUse(sym, loc) => addDependency(sym.loc, loc)
+    case SymUse.SigSymUse(sym, loc) => addDependency(sym.loc, loc)
+    case SymUse.StructFieldSymUse(sym, loc) => addDependency(sym.loc, loc)
+    case SymUse.TraitSymUse(sym, loc) => addDependency(sym.loc, loc)
+    case _ => ()
+  }
 
-      override def consumeCaseSymUse(symUse: SymUse.CaseSymUse): Unit = {
-        addDependency(symUse.sym.loc, symUse.loc)
-      }
+  /**
+   * Returns the dependencies of the given instance.
+   */
+  private def visitInstances(instances: List[TypedAst.Instance])(implicit sctx: SharedContext): Unit =
+    instances.foreach(instance => addDependency(instance.trt.sym.loc, instance.loc))
 
-      override def consumeDefSymUse(symUse: SymUse.DefSymUse): Unit = {
-        addDependency(symUse.sym.loc, symUse.loc)
-      }
+  private def visitSpec(spec: TypedAst.Spec)(implicit sctx: SharedContext): Unit = spec match {
+    case TypedAst.Spec(_, _, _, _, fparams, declaredScheme, retTpe, eff, tconstrs, econstrs) =>
+      visitType(retTpe)
+      visitScheme(declaredScheme)
+      visitType(eff)
+      fparams.foreach(visitFParam)
+      tconstrs.foreach(visitTraitConstraint)
+      econstrs.foreach(visitEqualityConstraint)
```

Comment:
I think you should visit the elements of each thing in the same order.

So here it would be fparams, declaredScheme, retTpe, and so on.